crates/matrix-sdk-appservice/tests/tests.rs

```diff
@@ -299,6 +304,105 @@ async fn test_appservice_on_sub_path() -> Result<()> {
     Ok(())
 }
 
+#[async_test]
+async fn test_receive_transaction() -> Result<()> {
+    tracing_subscriber::fmt().try_init().ok();
+    let json = vec![
+        Raw::new(&json!({
+            "content": {
+                "avatar_url": null,
+                "displayname": "Appservice",
+                "membership": "join"
+            },
+            "event_id": "$151800140479rdvjg:localhost",
+            "membership": "join",
+            "origin_server_ts": 151800140,
+            "sender": "@_appservice:localhost",
+            "state_key": "@_appservice:localhost",
+            "type": "m.room.member",
+            "room_id": "!coolplace:localhost",
+            "unsigned": {
+                "age": 2970366
+            }
+        }))?
+        .cast::<AnyRoomEvent>(),
+        Raw::new(&json!({
+            "content": {
+                "avatar_url": null,
+                "displayname": "Appservice",
+                "membership": "join"
+            },
+            "event_id": "$151800140491rfbja:localhost",
+            "membership": "join",
+            "origin_server_ts": 151800140,
+            "sender": "@_appservice:localhost",
+            "state_key": "@_appservice:localhost",
+            "type": "m.room.member",
+            "room_id": "!boringplace:localhost",
+            "unsigned": {
+                "age": 2970366
+            }
+        }))?
+        .cast::<AnyRoomEvent>(),
+        Raw::new(&json!({
+            "content": {
+                "avatar_url": null,
+                "displayname": "Alice",
+                "membership": "join"
+            },
+            "event_id": "$151800140517rfvjc:localhost",
+            "membership": "join",
+            "origin_server_ts": 151800140,
+            "sender": "@_appservice_alice:localhost",
+            "state_key": "@_appservice_alice:localhost",
+            "type": "m.room.member",
+            "room_id": "!coolplace:localhost",
+            "unsigned": {
+                "age": 2970366
+            }
+        }))?
+        .cast::<AnyRoomEvent>(),
+        Raw::new(&json!({
+            "content": {
+                "avatar_url": null,
+                "displayname": "Bob",
+                "membership": "invite"
+            },
+            "event_id": "$151800140594rfvjc:localhost",
+            "membership": "invite",
+            "origin_server_ts": 151800174,
+            "sender": "@_appservice_bob:localhost",
+            "state_key": "@_appservice_bob:localhost",
+            "type": "m.room.member",
+            "room_id": "!boringplace:localhost",
+            "unsigned": {
+                "age": 2970366
+            }
+        }))?
+        .cast::<AnyRoomEvent>(),
```

Comment:
I wasn't sure where to put the test json, as the needed `Vec<Raw>` form was slightly different than usual.

Reply:
Maybe into https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-test/src/appservice.rs?

Reply:
In most cases I prefer to put the JSON close to the test, even at the cost of some duplication. Otherwise it makes it harder to keep in mind which JSON we're using and what assumptions it carries with it.